packages/monaco/src/browser/monaco-workspace.ts

```diff
@@ -175,14 +182,34 @@ export class MonacoWorkspace implements lang.Workspace {
     @inject(ProblemManager)
     protected readonly problems: ProblemManager;
 
+    protected _workspaceFolders: WorkspaceFolder[];
+    get workspaceFolders(): WorkspaceFolder[] {
+        return this._workspaceFolders;
+    }
+
     @postConstruct()
     protected init(): void {
         this.workspaceService.roots.then(roots => {
-            const rootStat = roots[0];
-            if (rootStat) {
-                this._rootUri = rootStat.uri;
+            this._workspaceFolders = roots.map((dir, index) => this.toWorkspaceFolder(dir.uri, index));
+            if (this._workspaceFolders.length !== 0) {
+                this.resolveReady();
+            }
+        });
+
+        this.workspaceService.onWorkspaceChanged(async newRootDirs => {
+            const oldWorkspaceUris = this.workspaceFolders.map(folder => folder.uri.toString());
+            const newWorkspaceUris = newRootDirs.map(folder => folder.uri);
+            const added = newWorkspaceUris.filter(uri => oldWorkspaceUris.indexOf(uri) < 0).map((dir, index) => this.toWorkspaceFolder(dir, index));
+            const removed = oldWorkspaceUris.filter(uri => newWorkspaceUris.indexOf(uri) < 0).map((dir, index) => this.toWorkspaceFolder(dir, index));
+            this._workspaceFolders = newWorkspaceUris.map(this.toWorkspaceFolder);
+            if (this._workspaceFolders.length === 0) {
+                this._ready = new Promise<void>(resolve => {
```

Comment:
i don't think it is necessary. `roots` is a promised which resolved when a workspace is changed first time

Reply:
Ah, ok, so it is because of the race condition, between ready and `onWorkspaceChanged` event. I think if we construct code like in https://github.com/eclipse-theia/theia/pull/7182/files#r381844782 then it should not be necessary